src/reference.cr

```diff
@@ -159,8 +173,68 @@ class Reference
       false
     else
       hash[key] = true
-      value = yield
-      hash.delete(key)
+      begin
+        yield
+      ensure
+        hash.delete(key)
+      end
+      true
+    end
+  end
+
+  private def exec_recursive_outer(method)
```

Comment:
It is variant of `exec_recursive`, which jumps to outmost call when recursion is detected.

One of usage of this is `hash` calculation. For example, let `a = [b]` and `b = [b]`, it should be `a == b` because they are same structure, and so `a.hash == b.hash`. However recursion is detected at `a == a[0][0]` on `a`, but detected at `b == b[0]` on `b`, different recursion point produces different `hash` value and it causes a problem on storing to `Hash`. `exec_recursive_outer` is needed to prevent this.

For now, `hash` calculation is only use-case. But this method is called in some places.

Name of `exec_recurse_outer` is derived from Ruby's one, and I don't know why this is named. ([here](https://github.com/ruby/ruby/blob/eaf06f457af92505fb718a8b711f4d825f399b69/thread.c#L5105-L5112))

I think `exec_recursive_shortcircuit` is better (?)